README.md
# AzurePublicIPAddresses

[![Build Status](https://poshsecurity.visualstudio.com/PowerShell/_apis/build/status/AzurePublicIPAddresses?branchName=master)](https://poshsecurity.visualstudio.com/PowerShell/_build/latest?definitionId=36&branchName=master)

## Description

Gets the IP ranges associated with Azure regions in CIDR format for all public regions.

Authored by Kieran Jacobsen

## Installing

The easiest way to get AzurePublicIPAddresses is using the [PowerShell Gallery](https://powershellgallery.com/packages/AzurePublicIPAddresses/)!

### Inspecting the module

Best practice is that you inspect modules prior to installing them. You can do this by saving the module to a local path:

``` PowerShell
PS> Save-Module -Name AzurePublicIPAddresses -Path <path>
```

### Installing the module

Once you trust a module, you can install it using:

``` PowerShell
PS> Install-Module -Name AzurePublicIPAddresses
```

### Updating AzurePublicIPAddresses

Once installed from the PowerShell Gallery, you can update it using:

``` PowerShell
PS> Update-Module -Name AzurePublicIPAddresses
```

### Removing AzurePublicIPAddresses

To remove AzurePublicIPAddresses:

``` PowerShell
PS> Uninstall-Module -Name AzurePublicIPAddresses
```

## Contributing to AzurePublicIPAddresses

Interested in contributing? Read how you can [Contribute](contributing.md) to AzurePublicIPAddresses

This project is maintains a [Code of Conduct](code-of-conduct.md) that establishes how the project is governed and how everyone involved is expected to behave. You can report unacceptable behavior to [code@poshsecurity.com](mailto:code@poshsecurity.com).

## Release History

A detailed release history is contained in the [Change Log](CHANGELOG.md).

## License

AzurePublicIPAddresses is provided under the [MIT license](LICENSE.md).


CHANGELOG.md
# AzurePublicIPAddresses Release History

## 1.0.14 - 2019-04-22


### Added
* Support for UAE regions.
* Additional test cases.

## 1.0.11 - 2019-04-21


### Added
* Added support for South African, Chile and Brazil regions.
* New CI processs including GitVersion and YML build.


## 1.0.1 - 2019-03-24
## Added
* Includes support for Korean regions.


## 1.0 - 2018-05-01

### Fixed

* Handling for unknown regions.

### Added

* Added support for Australia Central, Australia Central 2, UK North, UK South 2, and placeholders North Europe 2 and East Europe.
* Added Readme, change log,  code of conduct, issue and pull request templates.
* Better tests including help content and more script analyser tests.

## 0.9 - 2018-02-04

### Added

* Support for France and Germany regions.
* Support for downloading the Germany region specific file.

### Changed

* Get-MicrosoftAzureDatacenterIPRangeFile parameters have changed. -ChinaRegion has been replaced with -Region China and -Region Germany.

## 0.8.3 - 2017-03-05

### Added

* Support for Korea regions.

## 0.8.2 - 2016-12-31

### Fixed

* CMDLetsToExport is now an empty array

### Added

* Support for US Central and US EAST2 EUAP regions.

## 0.8.1 - 2016-11-13

### Fixed

* Parameter validation for Get-MicrosoftAzureDatacenterIPRange and Get-MicrosoftAzureDatacenterIPRangeFile

### Changed

* Test cases for each Azure region.
* Comment based help has been improved.
* Better verbose output.


tests/Help.Tests.ps1
$script:ModuleName = 'AzurePublicIPAddresses'

# Removes all versions of the module from the session before importing
Get-Module $ModuleName | Remove-Module

$ModuleBase = Split-Path -Parent $MyInvocation.MyCommand.Path

# Get the list of functions we are not going to run tests against
$FunctionHelpTestExceptions = Get-Content -Path "$ModuleBase\Help.Exceptions.txt"

# For tests in .\Tests subdirectory
if ((Split-Path $ModuleBase -Leaf) -eq 'Tests') {
    $ModuleBase = Split-Path $ModuleBase -Parent
}

$Module = Import-Module $ModuleBase\$ModuleName.psd1 -PassThru -ErrorAction Stop
$commands = Get-Command -Module $module -CommandType Cmdlet, Function, Workflow  # Not alias

## When testing help, remember that help is cached at the beginning of each session.
## To test, restart session.

foreach ($command in $commands) {
    $commandName = $command.Name

    # Skip all functions that are on the exclusions list
    if ($script:FunctionHelpTestExceptions -contains $commandName) { continue } ## may not be correct check with a functionthat needs exceptions

    # The module-qualified command fails on Microsoft.PowerShell.Archive cmdlets
    $Help = Get-Help $commandName -ErrorAction SilentlyContinue

    Describe "Test help for $commandName" -Tag Help{

        # If help is not found, synopsis in auto-generated help is the syntax diagram
        It "should not be auto-generated" {
            $Help.Synopsis | Should Not BeLike '*`[`<CommonParameters`>`]*'
        }

        # Should be a description for every function
        It "gets description for $commandName" {
            $Help.Description | Should Not BeNullOrEmpty
        }

        # Should be at least one example
        It "gets example code from $commandName" {
            ($Help.Examples.Example | Select-Object -First 1).Code | Should Not BeNullOrEmpty
        }

        # Should be at least one example description
        It "gets example help from $commandName" {
            ($Help.Examples.Example.Remarks | Select-Object -First 1).Text | Should Not BeNullOrEmpty
        }

        Context "Test parameter help for $commandName" {

            $Common = 'Debug', 'ErrorAction', 'ErrorVariable', 'InformationAction', 'InformationVariable', 'OutBuffer', 'OutVariable',
            'PipelineVariable', 'Verbose', 'WarningAction', 'WarningVariable'

            $parameters = $command.ParameterSets.Parameters | Sort-Object -Property Name -Unique | Where-Object Name -notin $common
            $parameterNames = $parameters.Name
            $HelpParameterNames = $Help.Parameters.Parameter.Name | Sort-Object -Unique

            foreach ($parameter in $parameters) {
                $parameterName = $parameter.Name
                $parameterHelp = $Help.parameters.parameter | Where-Object Name -EQ $parameterName

                # Should be a description for every parameter
                It "gets help for parameter: $parameterName : in $commandName" {
                    $parameterHelp.Description.Text | Should Not BeNullOrEmpty
                }

                # Required value in Help should match IsMandatory property of parameter
                It "help for $parameterName parameter in $commandName has correct Mandatory value" {
                    $codeMandatory = $parameter.IsMandatory.toString()
                    $parameterHelp.Required | Should Be $codeMandatory
                }

                # Parameter type in Help should match code
                It "help for $commandName has correct parameter type for $parameterName" {
                    $codeType = $parameter.ParameterType.Name
                    # To avoid calling Trim method on a null object.
                    $helpType = if ($parameterHelp.parameterValue) { $parameterHelp.parameterValue.Trim() }
                    $helpType | Should be $codeType
                }
            }

            foreach ($helpParm in $HelpParameterNames) {
                # Shouldn't find extra parameters in help.
                It "finds help parameter in code: $helpParm" {
                    $helpParm -in $parameterNames | Should Be $true
                }
            }
        }
    }
}


tests/Project.Tests.ps1
$packages = get-package
if ($packages.Name  -contains "PSScriptAnalyzer") {
    #PSScriptAnalyzer is installed on the system
} else {
    Write-Output "Installing latest version of PSScriptAnalyzer"

    #install PSScriptAnalyzer
    Install-Package PSScriptAnalyzer -Force -Scope CurrentUser
}

$script:ModuleName = 'AzurePublicIPAddresses'

$ModuleBase = Split-Path -Parent $MyInvocation.MyCommand.Path

# Get the list of Pester Tests we are going to skip
$PesterTestExceptions = Get-Content -Path "$ModuleBase\Project.Exceptions.txt"

# For tests in .\Tests subdirectory
if ((Split-Path $ModuleBase -Leaf) -eq 'Tests') {
    $ModuleBase = Split-Path $ModuleBase -Parent
}

Describe "PSScriptAnalyzer rule-sets" -Tag Build , ScriptAnalyzer {

    $Rules = Get-ScriptAnalyzerRule
    $scripts = Get-ChildItem $ModuleBase -Include *.ps1, *.psm1, *.psd1 -Recurse | Where-Object fullname -notmatch 'classes'

    foreach ( $Script in $scripts )
    {
        Context "Script '$($script.FullName)'" {

            foreach ( $rule in $rules )
            {
                # Skip all rules that are on the exclusions list
                if ($PesterTestExceptions -contains $rule.RuleName) { continue }
                It "Rule [$rule]" {

                    (Invoke-ScriptAnalyzer -Path $script.FullName -IncludeRule $rule.RuleName ).Count | Should Be 0
                }
            }
        }
    }
}


Describe "General project validation: $moduleName" -Tags Build {
    BeforeAll {
        Get-Module $ModuleName | Remove-Module
    }
    It "Module '$moduleName' can import cleanly" {
        {Import-Module $ModuleBase\$ModuleName.psd1 -force } | Should Not Throw
    }
}

functions/Get-AzureRegionPublicIPAddresses.ps1
function Get-AzureRegionPublicIPAddresses {
    <#
    .SYNOPSIS
        Gets the IP ranges associated with Azure regions in CIDR format.
    .DESCRIPTION
        The Get-AzureRegionPublicIPAddresses cmdlet gets a list of subnets in CIDR format (eg 192.168.1.0/24). A specific region can be specified, otherwise this cmdlet will return all subnets from all regions.
        
        The cmdlet gets the information from the Microsoft Azure Datacenter IP Ranges file, this is updated weekly, and is available for download from: https://www.microsoft.com/en-us/download/details.aspx?id=41653
        
    .EXAMPLE
        C:\PS> <example usage>
        Explanation of what the example does
    .EXAMPLE
        C:\PS> <example usage>
        Explanation of what the example does
    .INPUTS
        Can take Azure region names from the pipeline.
    .OUTPUTS
        Outputs an array of subnets in CIDR format.
    #>
    
    [CmdletBinding()]
    param(       
        # Path to Microsoft Azure Datacenter IP Ranges file, this can be downloaded at: https://www.microsoft.com/en-us/download/details.aspx?id=41653
        [Parameter(Mandatory = $true, Position = 0)]
        [ValidateScript({Test-Path -Path $_})]
        [String]
        $AzurePublicIPXML,
        
        # Azure Datacenter/region
        [Parameter(Mandatory = $false, Position = 1, ValueFromPipeline = $true)]
        [ValidateSet(
                        'West Europe',
                        'East US',
                        'East US 2',
                        'US West',
                        'North Central US',
                        'North Europe',
                        'Central US',
                        'East Asia',
                        'Southeast Asia',
                        'South Central US',
                        'Japan West',
                        'Japan East',
                        'Brazil South',
                        'Australia East',
                        'Australia Southeast',
                        'Central India',
                        'West India',
                        'South India'
                    )]
        [String]
        $AzureRegion
    )
    
    begin {
        $PublicIPXML = [xml](Get-Content $AzurePublicIPXML)
    }
    
    process {
        # Are we selecting a specific region or returing all subnets?
        if ($PSBoundParameters.ContainsKey('AzureRegion'))
        {
            # Translate the friendly region name to the backend names
            switch ($AzureRegion)
            {
                'West Europe'           { $BackendRegionName = 'EuropeWest'             }
                'East US'               { $BackendRegionName = 'USEast'                 }
                'East US 2'             { $BackendRegionName = 'USEast2'                }
                'US West'               { $BackendRegionName = 'USWest'                 }
                'North Central US'      { $BackendRegionName = 'USNorth'                }
                'North Europe'          { $BackendRegionName = 'EuropeNorth'            }
                'Central US'            { $BackendRegionName = 'USCentral'              }
                'East Asia'             { $BackendRegionName = 'AsiaEast'               }
                'Southeast Asia'        { $BackendRegionName = 'AsiaSouthEast'          }
                'South Central US'      { $BackendRegionName = 'USSouth'                }
                'Japan West'            { $BackendRegionName = 'JapanWest'              }
                'Japan East'            { $BackendRegionName = 'JapanEast'              }
                'Brazil South'          { $BackendRegionName = 'BrazilSouth'            }
                'Australia East'        { $BackendRegionName = 'AustraliaEast'          }
                'Australia Southeast'   { $BackendRegionName = 'AustraliaSoutheast'     }
                'Central India'         { $BackendRegionName = 'IndiaCentral'           }
                'West India'            { $BackendRegionName = 'IndiaWest'              }
                'South India'           { $BackendRegionName = 'IndiaSouth'             }
            }
            
            Write-Verbose -Message "Backend region $BackendRegionName"
            Write-Output -InputObject ($PublicIPXML.AzurePublicIpAddresses.Region | Where-Object -FilterScript {$_.name -eq $BackendRegionName}).iprange.subnet   
        }
        else 
        {Write-Output -InputObject $PublicIPXML.AzurePublicIpAddresses.Region.iprange.subnet}
    }
}

functions/Get-MicrosoftAzureDatacenterIPRangeFile.ps1
#requires -Version 4.0

function Get-MicrosoftAzureDatacenterIPRangeFile
{
    <#
            .SYNOPSIS
            Downloads either the Microsoft Azure Datacenter IP Ranges file, or the Windows Azure Datacenter IP Ranges in China file. The file can be stored in memory or to the file system.

            .DESCRIPTION
            This CMDLet will download one of three files:
                Microsoft Azure Datacenter IP Ranges file (https://www.microsoft.com/en-us/download/details.aspx?id=41653)
                Windows Azure Datacenter IP Ranges in China (https://www.microsoft.com/en-us/download/details.aspx?id=42064)
                Windows Azure Datacenter IP Ranges in Germany (https://www.microsoft.com/en-us/download/details.aspx?id=54770)

            The region to be download is selected via the -Region parameter.

            It should be noted that this file is updated on a weekly basis, so if you save these file, then you should re-download them on a regular basis.

            This cmdlet makes use of Invoke-WebRequest.

            .EXAMPLE
            C:\PS> Get-MicrosoftAzureDatacenterIPRangeFile
            Returns an XML document from the Microsoft Azure Datacenter IP Ranges file

            .EXAMPLE
            C:\PS> Get-MicrosoftAzureDatacenterIPRangeFile -Path C:\Temp\AzureRanges.xml
            Dowloads the Microsoft Azure Datacenter IP Ranges file to C:\Temp\AzureRanges.xml

            .EXAMPLE
            C:\PS> Get-MicrosoftAzureDatacenterIPRangeFile -Region China
             Returns an XML document from the Microsoft Azure Datacenter IP Ranges in China file

            .EXAMPLE
            C:\PS> Get-MicrosoftAzureDatacenterIPRangeFile -Region China -Path C:\Temp\AzureRangesChina.xml
            Downloads the Microsoft Azure Datacenter IP Ranges in China file to C:\Temp\AzureRangesChina.xml

            .OUTPUTS
            XML document containing Azure Subnets, or
            Void if saving the file to the file system
    #>
    [CmdletBinding(DefaultParameterSetName = 'xml')]
    [OutputType([System.Xml.XmlDocument], ParameterSetName = 'path')]
    [OutputType([void], ParameterSetName = 'xml')]
    param(
        # Path where you want to save the XML file
        [Parameter(Mandatory        = $false,
                   ParameterSetName = 'path',
                   Position         = 0)]
        [ValidateNotNullOrEmpty()]
        [ValidateScript({Test-Path -Path (Split-Path -Path $_ -Parent) -PathType Container})]
        [String]
        $Path,

        # Specifies which region file to download, Standard, China or Germany.
        [Parameter(Mandatory = $false,
                   Position  = 1)]
        [ValidateNotNullOrEmpty()]
        [ValidateSet('Standard', 'China', 'Germany')]
        [string]
        $Region = 'Standard'
    )

    Write-Warning -Message '[Deprecation Notice] This file will be deprecated by June 30, 2020. Please switch to the JSON Azure Tags file or the Azure AZ module.'

    switch ($Region) {
        'China' {
            Write-Verbose -Message 'Downloading... Windows Azure Datacenter IP Ranges in China'
            $MicrosoftDownloadsURL = 'https://www.microsoft.com/en-us/download/confirmation.aspx?id=42064'
        }

        'Germany' {
            Write-Verbose -Message 'Downloading... Windows Azure Datacenter IP Ranges in Germany'
            $MicrosoftDownloadsURL = 'https://www.microsoft.com/en-us/download/confirmation.aspx?id=54770'
        }

        Default {
            Write-Verbose -Message 'Downloading... Microsoft Azure Datacenter IP Ranges'
            $MicrosoftDownloadsURL = 'https://www.microsoft.com/en-us/download/confirmation.aspx?id=41653'
        }
    }

    $DownloadPage = Invoke-WebRequest -UseBasicParsing -Uri $MicrosoftDownloadsURL
    $DownloadLink = ($DownloadPage.Links | Where-Object -FilterScript {$_.outerHTML -match 'Click here' -and $_.href -match '.xml'}).href[0]

    if ($PSCmdlet.ParameterSetName -eq 'path')
    {
        Write-Verbose -Message ('Saving file to {0}' -f $Path)
        Invoke-WebRequest -UseBasicParsing -Uri $DownloadLink -OutFile $Path
    }
    else
    {
        Write-Verbose -Message 'Downloading and creating XML object'
        $Request = Invoke-WebRequest -UseBasicParsing -Uri $DownloadLink
        $RequestXML = Select-Xml -Content $Request.toString() -XPath /
        $RequestXML.Node
    }
}


functions/Get-MicrosoftAzureDatacenterIPRange.ps1
#requires -Version 4.0
function Get-MicrosoftAzureDatacenterIPRange
{
    <#
            .SYNOPSIS
            Gets the IP ranges associated with Azure regions in CIDR format.

            .DESCRIPTION
            The Get-MicrosoftAzureDatacenterIPRange cmdlet gets a list of subnets in CIDR format (eg 192.168.1.0/24). You can get the ranges for a specific region, or get all subnets for all regions.

            There are three files that contain all the information that this CMDLet uses:
                Microsoft Azure Datacenter IP Ranges file (https://www.microsoft.com/en-us/download/details.aspx?id=41653)
                Windows Azure Datacenter IP Ranges in China (https://www.microsoft.com/en-us/download/details.aspx?id=42064)
                Windows Azure Datacenter IP Ranges in Germany (https://www.microsoft.com/en-us/download/details.aspx?id=54770)

            These files are updated each week.

            If the -path parameter is omitted, then the CMDLet will download both files and store them in memory (it doesn't save them to disk). The CMDLet will not download the file upon each execution,
            as it checks to see if the file has already been stored in memory.

            If you specify -path, the CMDLet will generate warnings that not all regions will be available. This is by design. I will look at handling both files saved to the file system in a later release.

            .EXAMPLE
            C:\PS> Get-MicrosoftAzureDatacenterIPRange -AzureRegion 'North Central US'
            Returns all of the subnets in the North Central US regions, will download the Microsoft Azure Datacenter IP Ranges file into memory

            .EXAMPLE
            C:\PS> Get-MicrosoftAzureDatacenterIPRange -AzureRegion @('North Central US', 'West India')
            Returns all of the subnets in the North Central US and West India regions, will download the Microsoft Azure Datacenter IP Ranges file into memory

            .EXAMPLE
            C:\PS> @('Japan West', 'Japan East') | Get-MicrosoftAzureDatacenterIPRange
            Returns all of the subnets in the Japan West and East regions, will download the Microsoft Azure Datacenter IP Ranges file into memory

            .EXAMPLE
            C:\PS> Get-MicrosoftAzureDatacenterIPRange -Path C:\Temp\AzureRanges.xml -AzureRegion 'North Central US'
            Returns all of the subnets in the North Central US region based on the specified file

            .EXAMPLE
            C:\PS> Get-MicrosoftAzureDatacenterIPRange
            Returns all of the subnets used by Azure, will download the Microsoft Azure Datacenter IP Ranges file into memory

            .INPUTS
            Can take Azure region names from the pipeline.

            .OUTPUTS
            Outputs objects containing each subnet and their region.
    #>
    [CmdletBinding()]
    [OutputType([PSCustomObject])]
    param(
        # Azure Datacenter/region
        [Parameter(Mandatory         = $false,
                   Position          = 0,
                   ValueFromPipeline = $true)]
        [ValidateSet(
            'Australia Central',
            'Australia Central 2',
            'Australia East',
            'Australia Southeast',
            'Brazil South',
            'Canada Central',
            'Canada East',
            'Central India',
            'Central US EUAP',
            'Central US',
            'China East',
            'China North',
            'East Asia',
            'East US 2 EUAP',
            'East US 2',
            'East US',
            'France Central',
            'France South',
            'Germany Central',
            'Germany Northeast',
            'Japan East',
            'Japan West',
            'Korea Central',
            'Korea South',
            'North Central US',
            'North Europe',
            'South Central US',
            'South India',
            'Southeast Asia',
            'UK North',
            'UK South',
            'UK South 2',
            'UK West',
            'West Central US',
            'West Europe',
            'West India',
            'West US 2',
            'West US',
            'South Africa North',
            'South Africa West',
            'UAE North',
            'UAE Central',
            'Germany North',
            'Germany West Central',
            'Norway East',
            'Norway West',
            'Switzerland North',
            'Switzerland West'
        )]
        [ValidateNotNullOrEmpty()]
        [String[]]
        $AzureRegion,

        # Path to Microsoft Azure Datacenter IP Ranges file
        [Parameter(Mandatory = $false,
                   Position  = 1)]
        [ValidateNotNullOrEmpty()]
        [ValidateScript({Test-Path -Path $_ -PathType Leaf})]
        [String]
        $Path
    )

    begin
    {
        if ($PSBoundParameters.ContainsKey('path'))
        {
            # Read the file
            $PublicIPXML = [xml](Get-Content -Path $Path)
            $Script:Regions = $PublicIPXML.AzurePublicIpAddresses.Region

            # Display warning stating that some regions may not be available
            Write-Warning -Message 'Using -Path may result in some regions being unavailable and returning no values.'
        }
        else
        {
            if ($null -eq $Script:Regions)
            {
                Write-Verbose -Message 'Fetching the standard region data file'
                $Script:Regions = (Get-MicrosoftAzureDatacenterIPRangeFile).AzurePublicIpAddresses.Region

                Write-Verbose -Message 'Fetching the China region file'
                $Script:Regions += (Get-MicrosoftAzureDatacenterIPRangeFile -Region 'China').AzurePublicIpAddresses.Region

                Write-Verbose -Message 'Fetching the Germany region file'
                $Script:Regions += (Get-MicrosoftAzureDatacenterIPRangeFile -Region 'Germany').AzurePublicIpAddresses.Region
            }
        }

        # Define the region mappings (What we see in the portal to the names in the file)
        $AzureRegions = @{
            'West Europe'          = 'EuropeWest'
            'East US'              = 'USEast'
            'East US 2'            = 'USEast2'
            'West US'              = 'USWest'
            'North Central US'     = 'USNorth'
            'North Europe'         = 'EuropeNorth'
            'Central US'           = 'USCentral'
            'East Asia'            = 'AsiaEast'
            'Southeast Asia'       = 'AsiaSouthEast'
            'South Central US'     = 'USSouth'
            'Japan West'           = 'JapanWest'
            'Japan East'           = 'JapanEast'
            'Brazil South'         = 'BrazilSouth'
            'Australia East'       = 'AustraliaEast'
            'Australia Southeast'  = 'AustraliaSoutheast'
            'Central India'        = 'IndiaCentral'
            'West India'           = 'IndiaWest'
            'South India'          = 'IndiaSouth'
            'Canada Central'       = 'CanadaCentral'
            'Canada East'          = 'CanadaEast'
            'West Central US'      = 'USWestCentral'
            'West US 2'            = 'USWest2'
            'UK South'             = 'UKSouth'
            'UK West'              = 'UKWest'
            'China North'          = 'chinanorth'
            'China East'           = 'chinaeast'
            'Central US EUAP'      = 'uscentraleuap'
            'East US 2 EUAP'       = 'useast2euap'
            'Korea South'          = 'koreasouth'
            'Korea Central'        = 'koreacentral'
            'France Central'       = 'francec'
            'France South'         = 'frances'
            'Germany Central'      = 'germanycentral'
            'Germany Northeast'    = 'germanynortheast'
            'Australia Central'    = 'australiac'
            'Australia Central 2'  = 'australiac2'
            'UK North'             = 'uknorth'
            'UK South 2'           = 'uksouth2'
            'South Africa North'   = 'southafrican'
            'South Africa West'    = 'southafricaw'
            'UAE North'            = 'uaen'
            'UAE Central'          = 'uaec'
            'Germany North'        = 'germanyn'
            'Germany West Central' = 'germanywc'
            'Norway East'          = 'norwaye'
            'Norway West'          = 'norwayw'
            'Switzerland North'    = 'switzerlandn'
            'Switzerland West'     = 'switzerlandw'
        }
    }

    process
    {
        # Are we selecting a specific region or returing all subnets?
        if ($PSBoundParameters.ContainsKey('AzureRegion'))
        {
            #We need to process this as we may have one or many regions specified in $AzureRegion
            foreach ($Region in $AzureRegion)
            {
                # Translate the friendly region name to the backend names
                $BackendRegionName = $AzureRegions[$Region]

                Write-Verbose -Message "Backend region: $BackendRegionName"
                $Subnets = ($Regions.where({$_.name -eq $BackendRegionName})).iprange.subnet
                foreach ($Subnet in $Subnets)
                {
                    $OutputObject = [PSCustomObject]@{
                        Region = $Region
                        Subnet = $Subnet
                    }
                    Write-Output -InputObject $OutputObject
                }
            }
        }
        else
        {
            foreach ($Region in $Regions)
            {
                $BackendRegionName = $Region.Name

                Write-Verbose -Message "Backend region: $BackedRegionName"

                # Translate each region name to something friendly
                $RegionName = $AzureRegions.GetEnumerator().Where({$_.Value -eq $BackendRegionName}).Name
                if ($null -eq $RegionName) {
                    Write-Warning -Message "Could not find Azure Region: $BackendRegionName"
                }

                Write-Verbose -Message "Azure Region Name: $RegionName"

                # Create the output object
                foreach ($Subnet in $Region.IpRange)
                {
                    $OutputObject = [PSCustomObject]@{
                        Region = $RegionName
                        Subnet = $Subnet.Subnet
                    }
                    Write-Output -InputObject $OutputObject
                }
            }
        }
    }
}
